clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -799,64 +804,94 @@ protected boolean onJoinPrepare(Timer timer, int generation, String memberId) {
-                exception = invokePartitionsLost(revokedPartitions);
-
+                        "indicating that consumer is in old state or no longer part of the group");
+                exception = Optional.ofNullable(invokePartitionsLost(partitions));
                 subscriptions.assignFromSubscribed(Collections.emptySet());
             }
         } else {
             switch (protocol) {
                 case EAGER:
-                    // revoke all partitions
-                    revokedPartitions.addAll(subscriptions.assignedPartitions());
-                    exception = invokePartitionsRevoked(revokedPartitions);
-
+                    exception = Optional.ofNullable(invokePartitionsRevoked(partitions));
                     subscriptions.assignFromSubscribed(Collections.emptySet());
 
                     break;
 
                 case COOPERATIVE:
-                    // only revoke those partitions that are not in the subscription any more.
                     Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
-                    revokedPartitions.addAll(ownedPartitions.stream()
-                        .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
-                        .collect(Collectors.toSet()));
-
-                    if (!revokedPartitions.isEmpty()) {
-                        exception = invokePartitionsRevoked(revokedPartitions);
-
-                        ownedPartitions.removeAll(revokedPartitions);
+                    if (!partitions.isEmpty()) {
+                        exception = Optional.ofNullable(invokePartitionsRevoked(partitions));
+                        ownedPartitions.removeAll(partitions);
                         subscriptions.assignFromSubscribed(ownedPartitions);
                     }
-
                     break;
             }
         }
 
-        isLeader = false;
-        subscriptions.resetGroupSubscription();
-        joinPrepareTimer = null;
-        autoCommitOffsetRequestFuture = null;
-        timer.update();
+        return exception;
+    }
 
-        if (exception != null) {
-            throw new KafkaException("User rebalance callback throws an error", exception);
-        }
-        return true;
+    private void markPartitionsUnconsumable(final Set<TopicPartition> partitions) {
+        // KAFKA-14196 for more detail, we pause the partition from consumption to prevent duplicated
+        //  data returned by the consumer poll loop.  Without pausing the partitions, the consumer will move forward
+        //  returning the data w/o committing them.  And the progress will be lost once the partition is revoked.
+        //  This only applies to autocommits, as we expect user to handle the offsets menually during the partition
+        //  revocation.
+        log.debug("Marking assigned partitions unconsumable: {}", partitions);
+        partitions.forEach(subscriptions::markUnconsumable);
```

Comment:
nit: how about we make `SubscriptionState.markUnconsumable` batched? Then we just need to grab the lock once.